### ISSUE TITLE
Achieving "idle while awaiting external event" behavior of Durable Functions, using DTFx

### ISSUE THREAD

**Issue by jplane** [Azure/durabletask#499]
TL/DR:

Given that Durable Function invocations in an Azure consumption plan [incur no resource costs when idling for external events](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-billing#awaiting-and-yielding-in-orchestrator-functions), can other code based on DTFx and hosted in a Functions consumption plan (but not itself implemented using Durable Functions) achieve the same effect?

Details:

I'm implementing a declarative state machine library built on DTFx:

https://github.com/jplane/StateChartsDotNet

...and includes a Functions-based hosting model:

https://github.com/jplane/StateChartsDotNet/tree/main/FunctionHost

In my implementation, state machines can wait for external events. I'm using TaskCompletionSource for that:

https://github.com/jplane/StateChartsDotNet/blob/main/DurableEngine/ExternalMessageQueue.cs

What else (if anything) is needed to achieve the "idle at zero" behavior of Durable Functions in a consumption plan?

Thanks for any pointers or help!

**Comment by jplane** [comment]
To answer my own question... after some adhoc testing it appears that await-ing a TaskCompetionSource is sufficient to achieve the consumption-based cost model.

I used the mechanisms described [here](https://docs.microsoft.com/en-us/azure/azure-functions/functions-consumption-costs#viewing-cost-related-data) to aggregate Function Execution Units over a few hours' time, during which my state machine app was mostly idle awaiting a TaskCompletionSource. By observation, no cost was incurred while the app was idle.

I'd still be grateful for confirmation from someone on the Functions team, if possible.  :-)

**Comment by cgillum** [comment]
Hmm...I would not expect this to be able to work, so I'm curious what kind of test you wrote and how you validated it. I expect there might need to be some clarifications. :)

In a nutshell, Azure Functions Consumption model bills you for the duration of a function execution. It's possible to cheat this by running threads in the background outside of the main execution thread, but that's cheating the system and creates a bunch of other problems since it doesn't look like real function executions to the Azure Functions infrastructure (unexpected process shutdowns, failure to scale-out, incorrect reporting, etc.).

What you're trying to do at a minimum requires you to implement an Azure Functions trigger that knows how to mark a function execution as started and stopped, as well as how to scale your app from 0 to 1 instance (and beyond). That trigger also needs to be supported by the Azure Functions Scale Controller so that it knows how to make scale decisions for your state machine. Unfortunately customer triggers are not supported in the Consumption plan and only the Azure Functions product team can build them (Elastic Premium plans have a bit more flexibility here, but the billing and works differently and "scale-to-zero" is not supported).

For transparency, I work on Durable Functions, DTFx, and also the Azure Functions scale platform. I've considered whether we should support a state-machine pattern in DF since it seems like a useful pattern. If I were to do it today, I'd probably create it directly on top of Durable Functions since that would be the easiest way to get something working with Azure Functions, similar to how we created Durable Entities in collaboration with MSR. Starting from DTFx would be possible, but also quite a bit more work. Feel free to ping me offline (I see you are also at MS :)) if you're interested in discussing any of this in more depth.
